classes/data/File.class.php

```diff
@@ -241,8 +241,18 @@ private function calculateEncryptedFileSize()
         if ($lastChunkPadding == 0) {
             $lastChunkPadding = 16;
         }
-            
-        return $this->size + ($chunksMinusOne * $echunkdiff) + $lastChunkPadding + 16;
+
+        switch( $this->transfer->key_version ) {
+            case CryptoAppConstants::v2018_importKey_deriveKey:
+            case CryptoAppConstants::v2017_digest_importKey:
+                return $this->size + ($chunksMinusOne * $echunkdiff) + $lastChunkPadding + 16;
```

Comment:
Not sure if your math is completely correct in light of this remark on Page 20:

  * The last chunk is sent, and considered the closing chunk if it is partial. If the last chunk is a full-sized chunk, then a closing chunk with zero bytes of content is added, with the chunk counter incremented once more.

If I have `$this->size=100` and `$upload_chunk_size=10`, I would get `$chunksMinusOne=9` which, if I understand your code correctly, should have been `10` instead.  The number of chunks should match the situation for `$this->size=101`, not `$this->size=99` as seems to be the case here.

Reply:
These two are the existing AES-CBC modes and I've left their chunk calculations intact.

Reply:
Understood that you leave the existing code intact.  The comment relates to the usefulness of the code somewhat higher up, where I could not add remarks.  Can you please convince me that when the last chunk is full (4 MB is what you mentioned as a chunk size) that there will be one more chunk, sized 0, so that the partial send is a proof that the file submission ends there?

Reply:
I dont think the web crypto api allows you to play around at the "16 byte low level chunk" size. Which we may perhaps refer to as AES-chunks. I would expect trying to encrypt a zero byte array of data on the web crypto api to produce undefined behavior on one or more browsers.

I have tested the code on 4mb and 4mb+1 byte sized files. The former results in a single filesender chunk and the later results in two filesender chunks. Note that the actual byte of the ciphered content is different as per these calculations.

What is your cryptographers opinion of adding a known pad to an exactly 4mb final filesender chunk of data? I could append a known cleartext to force the web crypto api to add a new chunk.

Reply:
I was playing around with this using a upload_chunk_size=`4*1024*1024`
One of my testing sample files is 30091892 bytes. I have been testing in both contiguious and chunked storage. The later is handy because FileSender will store each filesender chunk in a different file.

This results in 7 full sized filesender chunks and one partial chunk. Each chunk has 32 bytes of overhead added by the encryption. The first 7 will consume a fill 4mb of data each and the final chunk will have:
```
30091892 - (7*4*1024*1024)
731764
```

The on disk storage for this encrypted file requires 30092148 bytes. This makes sense when one calculates the padding for the 7 full and 1 partial chunk.

```
(4*1024*1024+32)*7 + 731764+32
30092148
```

I assume in the initial comment above in the part "The last chunk is sent, and considered the closing chunk if it is partial." you are talking about the chunks used by the AES algorithm and not the 4mb filesender chunks.

Reply:
> I dont think the web crypto api allows you to play around at the "16 byte low level chunk" size. Which we may perhaps refer to as AES-chunks. I would expect trying to encrypt a zero byte array of data on the web crypto api to produce undefined behavior on one or more browsers.

If you give me a link, I'll be happy to look for you.  What I also suspect is that you'll be allowed to supply arbitrary sequences of bytes, not just 16-byte multiples.  It would be quite an affront if you cannot pass 0 bytes through all (but then again, this is not a properly designed API for crypto).  If you cannot encrypt 0 bytes on all platforms we should look for another way, with security and all, to signal that the preceding 4MB-chunk was the last.

> I have tested the code on 4mb and 4mb+1 byte sized files. The former results in a single filesender chunk and the later results in two filesender chunks. Note that the actual byte of the ciphered content is different as per these calculations.

Hm, this is not what the report specified; both should produce 2 FS-chunks, the second of which encodes, a payload of 0 bytes, but also a signature to prove the end of transmission.

> What is your cryptographers opinion of adding a known pad to an exactly 4mb final filesender chunk of data? I could append a known cleartext to force the web crypto api to add a new chunk.

My opinion is that everything can work *provided* that you are able to parse the encrypted+signed content into an unambiguous conclusion that the case is different.

  * If you add a trailer, you can never be sure that it is indeed the last bit... the same bytes might occur inside, including at the end of a 4MB-chunk.  So your proposal isn't quite there yet.

  * What *should* work is sending 0 bytes, as specified.  I would be rather shocked if the WebCrypto API is too lame to do that well.  It gives the most consistent code, so I would prefer it.  Having said that:

  * What could work is not sending 4MB but 4MB+1 to indicate that it is the last part, and exactly 4MB in size.  All other payload sizes of end blocks would be <4MB.  As a result, payload sizes != 4MB indicate the end.  The extra byte could be ignored (or might be checked to have a certain value).  In all cases, the signature must be valid (which AES-GCM assures and results in success/failure from the decoding operation).

  * What could work, and would be really simple, is to use AEAD information to share the length.  This length is probably available out-of-bound, that is in headers or another non-secure form.  The AEAD adds security to such non-secure data.  Now you do need to check the length provided against that bit of data, but that's plain and simple and probably already part of what you are doing.

Feed problems with these options if you need me to generate more ;-)

Reply:
> Each chunk has 32 bytes of overhead added by the encryption.

Yes, that would be 16 bytes of unencrypted IV and 16 bytes GMAC/signature.

> I assume in the initial comment above in the part "The last chunk is sent, and considered the closing chunk if it is partial." you are talking about the chunks used by the AES algorithm and not the 4mb filesender chunks.

I am talking of the payload that is 4MB for any but the closing chunk.  I am looking for an invariant, namely != 4MB, and that can be enforced by adding a possible 0 byte chunk.

```
4MB-chunk --> AES --> IV + E(4MB + GMAC) --> 4MB + 32
4MB-chunk --> AES --> IV + E(4MB + GMAC) --> 4MB + 32
4MB-chunk --> AES --> IV + E(4MB + GMAC) --> 4MB + 32
3MB-chunk --> AES --> IV + E(3MB + GMAC) --> 3MB + 32
```

Alternatively, when sending 8MB, it would be (as proposed, and assuming the other options given above are not needed)

```
4MB-chunk --> AES --> IV + E(4MB + GMAC) --> 4MB + 32
4MB-chunk --> AES --> IV + E(4MB + GMAC) --> 4MB + 32
0MB-chunk --> AES --> IV + E(0MB + GMAC) --> 0MB + 32
```